pkg/collect/etcd.go

```diff
@@ -0,0 +1,375 @@
+func (c *etcdDebug) cleanup() {
+	if !c.ephemeral || c.pod == nil {
+		return
+	}
+
+	// delete the pod
+	klog.V(2).Infof("deleting etcd troubleshoot pod %s in namespace %s", c.pod.Name, c.pod.Namespace)
+	err := c.client.CoreV1().Pods(c.pod.Namespace).Delete(context.Background(), c.pod.Name, metav1.DeleteOptions{
+		GracePeriodSeconds: new(int64), // delete immediately
+	})
+	if err != nil {
+		klog.Errorf("failed to delete pod %s: %v", c.pod.Name, err)
+	}
+}
+
+// executeCommand exec into the pod and run the command
+// it returns the stdout, stderr and error if any of the command
+func (c *etcdDebug) executeCommand(command string) ([]byte, []byte, error) {
+
+	// split command into a slice of strings
+	// e.g. "etcdctl endpoint health" -> ["etcdctl", "endpoint", "health"]
+	cdmArgs := strings.Fields(command)
+	cdmArgs = append(cdmArgs, c.args...)
+	klog.V(2).Infof("executing command: %q in pod %q (namespace %q)", strings.Join(cdmArgs, " "), c.pod.Name, c.pod.Namespace)
+
+	req := c.client.CoreV1().RESTClient().Post().
+		Resource("pods").
+		Name(c.pod.Name).
+		Namespace(c.pod.Namespace).
+		SubResource("exec")
+
+	req.VersionedParams(&corev1.PodExecOptions{
+		Command: cdmArgs,
+		Stdin:   false,
+		Stdout:  true,
+		Stderr:  true,
+		TTY:     false,
+	}, scheme.ParameterCodec)
+
+	exec, err := remotecommand.NewSPDYExecutor(c.clientConfig, "POST", req.URL())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var stdout, stderr bytes.Buffer
+	err = exec.StreamWithContext(c.context, remotecommand.StreamOptions{
+		Stdout: &stdout,
+		Stderr: &stderr,
+	})
+
+	return stdout.Bytes(), stderr.Bytes(), err
+}
+
+// waitForPodReady waits until the etcd troubleshoot pod is running
+func (c *etcdDebug) waitForPodReady() error {
+	timeout := 60 * time.Second
+	ticker := time.NewTicker(1 * time.Second)
+
+	ctx, cancel := context.WithTimeout(c.context, timeout)
+	defer cancel()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return errors.New("timeout waiting for etcd troubleshooting pod to be running")
+		case <-ticker.C:
+			pod, err := c.client.CoreV1().Pods(c.pod.Namespace).Get(c.context, c.pod.Name, metav1.GetOptions{})
+			if err != nil {
+				return errors.Wrap(err, "failed to get etcd troubleshoot pod")
+			}
+			if pod.Status.Phase == corev1.PodRunning {
+				// ok, pod is running
+				return nil
+			}
+			klog.V(2).Infof("waiting for etcd troubleshoot pod %q to be running, current status: %q", c.pod.Name, pod.Status.Phase)
+		}
+	}
+}
+
+// generateFilenameFromCommand generates a filename from the command
+// e.g. "etcdctl endpoint health" -> "endpoint-health"
+func generateFilenameFromCommand(command string) string {
+	parts := strings.Fields(command)
+	if len(parts) == 0 {
+		return ""
+	}
+	return strings.Join(parts[1:], "-")
+}
+
+// getFullPath returns the full path to the file
+// e.g. "endpoint-health" -> "etcd/endpoint-health.txt"
+func getFullPath(fileName string) string {
+	return fmt.Sprintf("%s/%s.txt", etcdOutputDir, fileName)
```

Comment:
nit: Wouldn't this be better named as `<filename>.json`?

Reply:
that makes sense @banjoh, I've updated the code.